tools/perf/ucc_pt_benchmark.cc

```diff
@@ -40,51 +40,97 @@ ucc_pt_benchmark::ucc_pt_benchmark(ucc_pt_benchmark_config cfg,
     }
 }
 
+typedef struct test_args {
+    ucc_pt_comm *    comm;
+    ucc_coll_args_t *args;
+    int              nwarmup;
+    int              niter;
+} test_args_t;
+
+typedef struct thread_args {
+    test_args_t *            test_args;
+    std::chrono::nanoseconds time;
+    int                      thread_id;
+    ucc_status_t             st;
+} thread_args_t;
+
+void *run_single_test(void *arg);
+
 ucc_status_t ucc_pt_benchmark::run_bench() noexcept
 {
-    size_t min_count = coll->has_range() ? config.min_count : 1;
-    size_t max_count = coll->has_range() ? config.max_count : 1;
-    ucc_status_t st;
-    ucc_coll_args_t args;
-    std::chrono::nanoseconds time;
+    size_t                 min_count = coll->has_range() ? config.min_count : 1;
+    size_t                 max_count = coll->has_range() ? config.max_count : 1;
+    ucc_status_t           st;
+    ucc_coll_args_t        args;
+    test_args_t            test_args;
+    int                    i, n_threads = comm->get_n_threads();
+    std::vector<pthread_t> threads;
+    threads.resize(n_threads);
+    std::vector<thread_args_t> threads_args;
+    threads_args.resize(n_threads);
 
     print_header();
     for (size_t cnt = min_count; cnt <= max_count; cnt *= 2) {
         size_t coll_size = cnt * ucc_dt_size(config.dt);
-        int iter = config.n_iter_small;
-        int warmup = config.n_warmup_small;
+        int    iter      = config.n_iter_small;
+        int    warmup    = config.n_warmup_small;
         if (coll_size >= config.large_thresh) {
-            iter = config.n_iter_large;
+            iter   = config.n_iter_large;
             warmup = config.n_warmup_large;
         }
         UCCCHECK_GOTO(coll->init_coll_args(cnt, args), exit_err, st);
-        UCCCHECK_GOTO(run_single_test(args, warmup, iter, time), free_coll, st);
+        test_args = (test_args_t){comm, &args, warmup, iter};
+        if (n_threads > 1) {
+            for (i = 0; i < n_threads; i++) {
+                threads_args.at(i).thread_id = i;
+                threads_args.at(i).test_args = &test_args;
```

Comment:
is it ok if you use same buffer for all threads? I know that we don't check correctness, but are there any other side-effects e.g. registration?